File: Day2/1_client-side-build/src/4_iterators/Queue.cs
using System;
using System.Collections;
using System.Collections.Generic;

public class Queue<T> : IEnumerable<T>
{
    private T[] _items;
    private int _front;
    private int _rear;
    private int _count;
    private int _capacity;

    public Queue(int capacity)
    {
        _capacity = capacity;
        _items = new T[_capacity];
        _front = 0;
        _rear = -1;
        _count = 0;
    }

    public void Enqueue(T item)
    {
        if (_count == _capacity)
        {
            throw new InvalidOperationException("Queue is full");
        }
        _rear = (_rear + 1) % _capacity;
        _items[_rear] = item;
        _count++;
    }

    public T Dequeue()
    {
        if (IsEmpty)
        {
            throw new InvalidOperationException("Queue is empty");
        }
        T item = _items[_front];
        _front = (_front + 1) % _capacity;
        _count--;
        return item;
    }

    public bool IsEmpty => _count == 0;

    public IEnumerator<T> GetEnumerator()
    {
        int current = _front;
        int counter = 0;

        while (counter < _count)
        {
            yield return _items[current];
            current = (current + 1) % _capacity;
            counter++;
        }
    }

    IEnumerator IEnumerable.GetEnumerator()
    {
        return GetEnumerator();
    }
}

File: Day2/1_client-side-build/src/4_iterators/Queue.java
import java.util.Iterator;
import java.util.NoSuchElementException;

public class Queue<T> implements Iterable<T> {
    private T[] items;
    private int front, rear, count, capacity;

    @SuppressWarnings("unchecked")
    public Queue(int capacity) {
        items = (T[]) new Object[capacity];
        this.capacity = capacity;
        front = 0;
        rear = -1;
        count = 0;
    }

    public void enqueue(T item) {
        if (count == capacity) {
            throw new IllegalStateException("Queue is full");
        }
        rear = (rear + 1) % capacity;
        items[rear] = item;
        count++;
    }

    public T dequeue() {
        if (isEmpty()) {
            throw new IllegalStateException("Queue is empty");
        }
        T item = items[front];
        front = (front + 1) % capacity;
        count--;
        return item;
    }

    public boolean isEmpty() {
        return count == 0;
    }

    @Override
    public Iterator<T> iterator() {
        return new Iterator<T>() {
            private int currentIndex = front;
            private int itemsReturned = 0;

            @Override
            public boolean hasNext() {
                return itemsReturned < count;
            }

            @Override
            public T next() {
                if (!hasNext()) {
                    throw new NoSuchElementException();
                }
                T item = items[currentIndex];
                currentIndex = (currentIndex + 1) % capacity;
                itemsReturned++;
                return item;
            }
        };
    }
}
